features/cellular/framework/AT/ATHandler.cpp

```diff
@@ -337,48 +335,41 @@ void ATHandler::rewind_buffer()
     }
 }
 
-// we are always expecting to receive something so there is wait timeout
-void ATHandler::fill_buffer()
+bool ATHandler::fill_buffer()
 {
     tr_debug("%s", __func__);
     // Reset buffer when full
     if (sizeof(_recv_buff) == _recv_len) {
         reset_buffer();
     }
 
-    Timer timer;
-    timer.start();
-    do {
-       ssize_t len = _fileHandle->read(_recv_buff + _recv_len, sizeof(_recv_buff) - _recv_len);
-        if (len > 0) {
-            _recv_len += len;
-           at_debug("\n----------readable----------: %d\n", _recv_len);
-           for (size_t i = _recv_pos; i < _recv_len; i++) {
-               at_debug("%c", _recv_buff[i]);
-           }
-           at_debug("\n----------readable----------\n");
-           return;
-       } else if (len != -EAGAIN && len != 0) {
-           tr_warn("%s error: %d while reading", __func__, len);
-           break;
-       }
-#ifdef MBED_CONF_RTOS_PRESENT
-        rtos::Thread::yield();
-#endif
-    } while ((uint32_t)timer.read_ms() < _at_timeout);
+    int timeout = (_start_time + _at_timeout) - rtos::Kernel::get_ms_count();
+    if (timeout >= 0) {
```

Comment:
@kjbracey-arm suggests to still attempt one poll with 0 timeout even if the timeout expired. Same in write().

Reply:
Yes, you may have overshot your precise time, presumably due to CPU load or scheduling or debugger interference, but if you are already in your own processing context you may as well have one last immediate check before you go. No need to say "timeout" if the data is actually already there.

This is consistent with what other APIs that take "until" times do, like `Semaphore::wait_until`. 

    uint64_t now = Kernel::get_ms_count();

    if (now >= millisec) {
        return wait(0);
    } else if (millisec - now >= osWaitForever) {
        // API permits early return
        return wait(osWaitForever - 1);
    } else {
        return wait(millisec - now);
    }

You might also want to handle the "target time is too big for int32_t" case. So:

    if (now >= _start_time + _at_timeout) {
       timeout = 0;
    } else if ( _start_time + _at_timeout - now > INT_MAX) {
       timeout = INT_MAX;
    } else {
       timeout = _start_time + _at_timeout - now;
    }

Unfortunate that there's no `poll_until` that does that for you. I possibly should add it; I didn't because it doesn't exist in any POSIX or C/C++ standard.